File: lemmy_client.py
from pprint import pprint
import requests

# Simple Lemmy client to fetch latest posts from a given instance
# Usage: python lemmy_client.py <instance_url> [community] [limit]
# EX: python lemmy_client.py https://lemmy.ml programming 5

import sys

def fetch_posts(instance_url, community=None, limit=10):
    """
    Fetch latest posts from Lemmy instance.
    If community is None, fetch site-wide posts.
    """
    base = instance_url.rstrip('/')
    url = f"{base}/api/v3/post/list"
    params = {
        'limit': limit,
        # 'sort': 'New',  # other options: Hot, Top, etc.
    }
    if community:
        params['community_name'] = community

    resp = requests.get(url, params=params)
    resp.raise_for_status()
    data = resp.json()
    return data.get('posts', [])


def main():
    if len(sys.argv) < 2:
        print("Usage: python lemmy_client.py <instance_url> [community] [limit]")
        sys.exit(1)

    instance = sys.argv[1]
    community = None
    limit = 10
    if len(sys.argv) >= 3:
        community = sys.argv[2]
    if len(sys.argv) == 4:
        try:
            limit = int(sys.argv[3])
        except ValueError:
            pass

    posts = fetch_posts(instance, community, limit)
    for post in posts:
        p = post['post']
        pprint(p)
        #print(f"{p['name']} \n  by {p['author']} - {p['url']}\n")

if __name__ == '__main__':
    main()

File: node/public/clients/lemmy_client.js
// lemmy_client.js
// Simple Lemmy client to fetch latest posts from a given instance
// Usage: node lemmy_client.js <instance_url> [community] [limit]

// EX: node lemmy_client.js https://lemmy.ml programming 5




/**
 * Fetch latest posts from a Lemmy instance.
 * @param {string} instanceUrl - Base URL of the Lemmy instance (e.g. https://lemmy.ml)
 * @param {string|null} community - Community name to filter by (e.g. 'programming'), or null for site-wide.
 * @param {number} limit - Number of posts to fetch (default: 10)
 * @returns {Promise<Array<Object>>} - Resolves to an array of post objects
 */
async function fetchPosts(instanceUrl, community = null, limit = 10) {
  const base = instanceUrl.replace(/\/+$/, '');
  const url = `${base}/api/v3/post/list`;
  const params = { limit };
  if (community) params.community_name = community;

  const response = await axios.get(url, { params });
  return response.data.posts || [];
}

/**
 * Print posts in readable form.
 * @param {Array<Object>} posts
 */
function printPosts(posts) {
  posts.forEach(({ post }) => {
    console.log('Title :', post.name);
    console.log('Author:', post.author);
    console.log('URL   :', post.url);
    console.log(post);
    console.log('---');
  });
}

/**
 * CLI entry point
 */
async function main() {
  const [,, instance, communityArg, limitArg] = process.argv;
  if (!instance) {
    console.error('Usage: node lemmy_client.js <instance_url> [community] [limit]');
    process.exit(1);
  }
  const community = communityArg || null;
  const limit = Number(limitArg) || 10;

  try {
    const posts = await fetchPosts(instance, community, limit);
    if (posts.length === 0) {
      console.log('No posts found.');
    } else {
      printPosts(posts);
    }
  } catch (error) {
    console.error('Error fetching posts:', error.message);
  }
}

// if (require.main === module) {
//   main();
// }

// module.exports = { fetchPosts, printPosts };
